slack/web/urllib_client.py

```diff
@@ -0,0 +1,266 @@
+from slack.web import get_user_agent, show_2020_01_deprecation, convert_bool_to_0_or_1
+from slack.web.slack_response import SlackResponse
+
+
+class HttpErrorResponse:
+    pass
+
+
+class UrllibWebClient:
+    logger = logging.getLogger(__name__)
+
+    def __init__(
+        self,
+        *,
+        token: str = None,
+        default_headers: Dict[str, str] = {},
+        # Not type here to avoid ImportError: cannot import name 'WebClient' from partially initialized module
+        # 'slack.web.client' (most likely due to a circular import
+        web_client=None,
+    ):
+        """urllib-based API client.
+
+        :param token: Slack API Token (either bot token or user token)
+        :param default_headers: request headers to add to all requests
+        :param web_client: WebClient instance for pagination
+        """
+        self.token = token
+        self.default_headers = default_headers
+        self.web_client = web_client
+
+    def api_call(
+        self,
+        *,
+        token: str = None,
+        url: str,
+        query_params: Dict[str, str] = {},
+        json_body: Dict = {},
+        body_params: Dict[str, str] = {},
+        files: Dict[str, io.BytesIO] = {},
+        additional_headers: Dict[str, str] = {},
+    ) -> SlackResponse:
+        """Performs a Slack API request and returns the result.
+
+        :param token: Slack API Token (either bot token or user token)
+        :param url: a complete URL (e.g., https://www.slack.com/api/chat.postMessage)
+        :param query_params: query string
+        :param json_body: json data structure (it's still a dict at this point),
+            if you give this argument, body_params and files will be skipped
+        :param body_params: form params
+        :param files: files to upload
+        :param additional_headers: request headers to append
+        :return: API response
+        """
+
+        show_2020_01_deprecation(self._to_api_method(url))
+
+        files_to_close: List[BinaryIO] = []
+        try:
+            # True/False -> "1"/"0"
+            query_params = convert_bool_to_0_or_1(query_params)
+            body_params = convert_bool_to_0_or_1(body_params)
+
+            if self.logger.level <= logging.DEBUG:
+                self.logger.debug(
+                    f"Slack API Request - url: {url}, "
+                    f"query_params: {query_params}, "
+                    f"json_body: {json_body}, "
+                    f"body_params: {body_params}, "
+                    f"files: {files}, "
+                    f"additional_headers: {additional_headers}"
+                )
+
+            request_data = {}
+            if files:
+                if body_params:
+                    for k, v in body_params.items():
+                        request_data.update({k: v})
+
+                for k, v in files.items():
+                    if isinstance(v, str):
+                        f: BinaryIO = open(v.encode("ascii", "ignore"), "rb")
```

Comment:
It seems we need to apply the change by #673 here

Reply:
done by https://github.com/slackapi/python-slackclient/pull/662/commits/ece7a12af8c36b571b131d656c418039414542b1